From 95fd9f53e1b78b3c1e1447bc75fdfc4a0000f592 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Mon, 11 Aug 2025 08:37:24 +0200
Subject: [PATCH 911/954] fix(stm32mp2): prevent panic when external oscillator
 is absent

To deactivate the external oscillator, the frequency must be set to 0
in the device tree.
The frequency check for 0 was implemented in clk_stm32_osc_gate_enable()
and clk_stm32_osc_gate_disable(), but missing in
clk_stm32_osc_gate_is_enabled().
Since clk_stm32_osc_gate_is_enabled() is called by clk_enable()
to verify if the clock is actually enabled, this fix prevents a panic
when the oscillator is not present.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 0d03a5d90af3bb40f1ac57d93ff2b795ef1bfcad fix(stm32mp2): prevent panic when external oscillator is absent]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I90dc671f39bd46d0db19d7532aee9ec7b449ba9d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/481607
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/st/clk/clk-stm32mp2.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/st/clk/clk-stm32mp2.c b/drivers/st/clk/clk-stm32mp2.c
index 9c93db578e..6552161fcf 100644
--- a/drivers/st/clk/clk-stm32mp2.c
+++ b/drivers/st/clk/clk-stm32mp2.c
@@ -846,6 +846,10 @@ static bool clk_stm32_osc_gate_is_enabled(struct stm32_clk_priv *priv, int id)
 {
 	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(priv, id);
 
+	if (osc_data->frequency == 0UL) {
+		return true;
+	}
+
 	return _clk_stm32_gate_is_enabled(priv, osc_data->gate_id);
 
 }
-- 
2.53.0

